File: app/src/main/java/com/quran/labs/androidquran/presenter/bookmark/RecentPagePresenter.java
package com.quran.labs.androidquran.presenter.bookmark;

import com.quran.labs.androidquran.data.Constants;
import com.quran.labs.androidquran.di.ActivityScope;
import com.quran.labs.androidquran.model.bookmark.RecentPageModel;
import com.quran.labs.androidquran.presenter.Presenter;
import com.quran.labs.androidquran.ui.PagerActivity;

import javax.inject.Inject;

import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;

@ActivityScope
public class RecentPagePresenter implements Presenter<PagerActivity> {
  private final RecentPageModel model;

  private int lastPage;
  private int minimumPage;
  private int maximumPage;
  private Disposable disposable;

  @Inject
  RecentPagePresenter(RecentPageModel model) {
    this.model = model;
  }

  private void onPageChanged(int page) {
    model.updateLatestPage(page);

    lastPage = page;
    if (minimumPage == Constants.NO_PAGE) {
      minimumPage = page;
      maximumPage = page;
    } else if (page < minimumPage) {
      minimumPage = page;
    } else if (page > maximumPage) {
      maximumPage = page;
    }
  }

  public void onJump() {
    saveAndReset();
  }

  @Override
  public void bind(PagerActivity what) {
    minimumPage = Constants.NO_PAGE;
    maximumPage = Constants.NO_PAGE;
    lastPage = Constants.NO_PAGE;

    disposable = what.getViewPagerObservable()
        .subscribeWith(new DisposableObserver<Integer>() {
          @Override
          public void onNext(Integer value) {
            onPageChanged(value);
          }

          @Override
          public void onError(Throwable e) {
          }

          @Override
          public void onComplete() {
          }
        });
  }

  @Override
  public void unbind(PagerActivity what) {
    disposable.dispose();
    saveAndReset();
  }

  private void saveAndReset() {
    if (minimumPage != Constants.NO_PAGE || maximumPage != Constants.NO_PAGE) {
      model.persistLatestPage(minimumPage, maximumPage, lastPage);

      minimumPage = Constants.NO_PAGE;
      maximumPage = Constants.NO_PAGE;
    }
    lastPage = Constants.NO_PAGE;
  }
}

File: app/src/main/java/com/quran/labs/androidquran/presenter/bookmark/RecentPagePresenter.kt
package com.quran.labs.androidquran.presenter.bookmark

import com.quran.labs.androidquran.data.Constants
import com.quran.data.di.ActivityScope
import com.quran.labs.androidquran.model.bookmark.RecentPageModel
import com.quran.labs.androidquran.presenter.Presenter
import com.quran.labs.androidquran.ui.PagerActivity
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.observers.DisposableObserver
import javax.inject.Inject

@ActivityScope
class RecentPagePresenter @Inject constructor(private val model: RecentPageModel) : Presenter<PagerActivity> {

  private var lastPage = 0
  private var minimumPage = 0
  private var maximumPage = 0
  private var disposable: Disposable? = null

  private fun onPageChanged(page: Int) {
    model.updateLatestPage(page)
    lastPage = page
    when {
        minimumPage == Constants.NO_PAGE -> {
          minimumPage = page
          maximumPage = page
        }
        page < minimumPage -> {
          minimumPage = page
        }
        page > maximumPage -> {
          maximumPage = page
        }
    }
  }

  fun onJump() {
    saveAndReset()
  }

  override fun bind(what: PagerActivity) {
    minimumPage = Constants.NO_PAGE
    maximumPage = Constants.NO_PAGE
    lastPage = Constants.NO_PAGE
    disposable = what.viewPagerObservable
      .subscribeWith(object : DisposableObserver<Int>() {
        override fun onNext(value: Int) {
          onPageChanged(value)
        }

        override fun onError(e: Throwable) {}
        override fun onComplete() {}
      })
  }

  override fun unbind(what: PagerActivity) {
    disposable?.dispose()
    saveAndReset()
  }

  private fun saveAndReset() {
    if (minimumPage != Constants.NO_PAGE || maximumPage != Constants.NO_PAGE) {
      model.persistLatestPage(minimumPage, maximumPage, lastPage)
      minimumPage = Constants.NO_PAGE
      maximumPage = Constants.NO_PAGE
    }
    lastPage = Constants.NO_PAGE
  }
}
